backend/app/rest/proxy/image.go

```diff
@@ -105,36 +145,78 @@ func (p Image) Handler(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 	}
-	if _, e := io.Copy(w, resp.Body); e != nil {
-		log.Printf("[WARN] can't copy image stream, %s", e)
+
+	w.Header().Add("Content-Type", "image/*")
+	_, err = io.Copy(w, imgReader)
+	if err != nil {
+		log.Printf("[WARN] can't copy image stream, %s", err)
 	}
 }
 
-// extract gets all non-https images and return list of src
-func (p Image) extract(commentHTML string) ([]string, error) {
-	doc, err := goquery.NewDocumentFromReader(strings.NewReader(commentHTML))
+// cache image from provided Reader using given ID
+func (p Image) cacheImage(r io.Reader, imgID string) {
+	id, err := p.ImageService.SaveWithID(imgID, r)
 	if err != nil {
-		return nil, errors.Wrap(err, "can't create document")
+		log.Printf("[WARN] unable to save image to the storage: %+v", err)
 	}
-	result := []string{}
-	doc.Find("img").Each(func(i int, s *goquery.Selection) {
-		if im, ok := s.Attr("src"); ok {
-			if strings.HasPrefix(im, "http://") {
-				result = append(result, im)
-			}
+	// In the future we can do something smarter than just committing everything (eg, some kind of LFU/LRU)
+	if err := p.ImageService.Commit(id); err != nil {
+		log.Printf("[WARN] unable to commit image %s", imgID)
+	}
+}
+
+// download an image. Returns a Reader which has to be closed by a caller
+func (p Image) downloadImage(ctx context.Context, imgURL string) (io.ReadCloser, error) {
+	log.Printf("[DEBUG] downloading image %s", imgURL)
+
+	timeout := 60 * time.Second // default
+	if p.Timeout > 0 {
+		timeout = p.Timeout
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	client := http.Client{Timeout: 30 * time.Second}
+	var resp *http.Response
+	err := repeater.NewDefault(5, time.Second).Do(ctx, func() error {
+		var e error
+		req, e := http.NewRequest("GET", imgURL, nil)
+		if e != nil {
+			return errors.Wrapf(e, "failed to make request for %s", imgURL)
 		}
+		resp, e = client.Do(req.WithContext(ctx))
+		return e
 	})
-	return result, nil
-}
+	if err != nil {
+		log.Print(err.Error())
```

Comment:
I think error should be printed with `[ERROR]` prefix or something like that, and some explanation where this error comes from.

Reply:
Fixed in #580.